### PR TITLE
List breaches

### PR DESCRIPTION
**Pull request by Vinnl**
# References: 
Jira: MNTOR-1343 and MNTOR-1344
Figma: https://www.figma.com/file/4Cw0BfI1iqu0HzGkxPmsdU/Breach-Detail-Page?node-id=1-5385&t=ulhnif4UYxXluY85-4

# Description

@rhelmer Most of the changes here are yours - I'm not sure if there were final touches you wanted to do, but if so, feel free to create a PR with just your changes, and I'll rebase on that.

This adds back the breach overview and breach details page to the refresh. Note that I did not add the background image on the overview - see [this Figma comment](https://www.figma.com/file/4Cw0BfI1iqu0HzGkxPmsdU?node-id=1:5385#395157513).

# Screenshot (if applicable)

image

image

# How to test

Click the "Breaches" link in the header or footer. Search of a breached company, then click the card.

# Checklist (Definition of Done)
- [x] Localization strings (if needed) have been added. - I think so, but possibly all strings have been given a do-over. I'm still checking this.
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [ ] I've added a unit test to test for potential regressions of this bug. - N/A, I think, lacking UI testing infra
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: locales/en/app.ftl [hunk 1]
```diff
@@ -400,7 +404,8 @@ fb-comp-and-others =
 no-other-breaches-found = No other breaches found from a basic search.
 
 no-results-blurb = Sorry, that breach is not in our database.
-all-breaches-headline = All breaches in { -product-name }
+all-breaches-headline-2 = All breaches detected by { -product-name }
```
Comment (flodolo) [inline locales/en/app.ftl:407]:
```suggestion
all-breaches-headline-2 = All breaches detected by { -brand-fx-monitor }
```

**Review by flodolo — CHANGES_REQUESTED** [review]

File: locales/en/app.ftl (hunk 1 above)
Comment (Vinnl) [reply locales/en/app.ftl]: 3992e326

File: src/utils/breach-logo.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * @param {unknkown} breach
+ * @param {Map<string, string>} logos Map of URLs to logos indexed by the domain name of the respective company
+ * @returns {string} HTML for a breach logo (either an `img`, or a `span.breach-logo` containing the breached company's first letter)
+ */
+export function getBreachLogo (breach, logos) {
+  const logo = logos.has(breach.Domain)
+    ? `<img src='${logos.get(breach.Domain)}' alt='' class='breach-logo' height='32' />`
```
Comment (toufali) [inline src/utils/breach-logo.js:12]: Cool that this can be repurposed for both instances!  

For the all-breaches instance, the Monitor logo in the header (and other loads like fonts) are noticeably delayed for me because of all the image requests.  I think a `loading="lazy"` attribute in the image tag would resolve that.

File: src/views/partials/all-breaches.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+</main>
+`
+
+function formatList (list) {
+  if (typeof Intl.ListFormat === 'undefined') {
+    return list.join(', ')
+  }
+
+  return (new Intl.ListFormat(getLocale(), { type: 'unit', style: 'short' })).format(list)
+}
```
Comment (toufali) [inline src/views/partials/all-breaches.js:85]: Nice!

File: src/utils/breach-logo.js
Comment (toufali) [inline src/utils/breach-logo.js:1]: Nice refactor into this module!

**Review by toufali — CHANGES_REQUESTED** [review]
Great job getting this together in rapid fashion!  I noted one blocker and a few other things that are less block-y, but somewhat important to me 😊

File: src/views/guestLayout.js [hunk 4]
```diff
@@ -61,6 +62,7 @@ const guestLayout = data => `
         <img src='/images/moz-logo-1color-white-rgb-01.svg' width='100' height='29' loading='lazy' alt='${getMessage('mozilla')}'>
       </a>
       <menu>
+        <li><a href='/breaches'>${getMessage('breaches')}</a></li>
```
[the remark this answers is not in the document]
Comment (toufali) [reply src/views/guestLayout.js:65]: Update, I posed this question to UX in Slack

File: src/client/css/global.css [hunk 5]
```diff
@@ -113,6 +113,27 @@ body > header .nav-wrapper {
...
+
+body > header menu a:hover {
+  text-decoration: underline;
+}
+
+@media (max-width: 480px) {
+  body > header menu a:not(.button) {
+    display: none;
+  }
+}
```
[the remark this answers is not in the document]
Comment (toufali) [reply src/client/css/global.css:135]: Update, I posed this question to UX in Slack, thanks to urgent prodding by @pdehaan

File: locales/en/app.ftl [hunk 6]
```diff
@@ -79,6 +79,9 @@ compromised-accounts = Compromised accounts:
 # compromised-data = the kind of user data exposed to hackers in data breach.
 compromised-data = Compromised data:
 
+# the kind of user data exposed to hackers in data breach.
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply locales/en/app.ftl]: I moved them to the V2 section in 068b0e73f.

I was hesitant about moving them to a new file already, since I'm not sure if that will need alignment with the Pontoon folks, so I left that for now.

File: src/client/js/search.js
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/client/js/search.js:1]: Moved it into `all-breaches.js`: b247e4c90

And fixed that darn error in ef4802a69 (looking forward to code splitting....)

File: src/client/css/partials/allBreachesPartial.css
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/client/css/partials/allBreachesPartial.css:1]: There's a lot of indirection going on, so it's a bit hard to follow, but if I understand it correctly, we automatically load CSS files whose file name, minus `.css`, matches the name of the function rendering the partial, so we can't use dashes there.

File: src/client/js/search.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+document.addEventListener('DOMContentLoaded', () => {
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/client/js/search.js]: I would assume so too - removed in 086ec77f9, but @rhelmer, let me know if that's a mistake.

File: src/utils/breach-resolution.js [hunk 8]
```diff
@@ -51,7 +51,7 @@ const breachResolutionDataTypes = {
   },
   [BreachDataTypes.CreditCard]: {
     priority: 4,
-    header: 'breach-checklist-cc-header',
+    header: 'q',
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/utils/breach-resolution.js]: @rhelmer I'm assuming [this](https://github.com/mozilla/blurts-server/commit/89beb67c1670daa9d684b5a6a3ec24f9cbd4f7ff#diff-004cb721b5ad3b7696a4e760d9732dab0cbd137382f201918e9534c8aac0e9cc) was a mistake, so I've reverted this in 754522a0a. Nice catch Amri!

File: src/views/guestLayout.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/views/guestLayout.js]: As per Mel's answer, updated in 38c9c3268.

**Review by flodolo — APPROVED** [review]
Strings look good, please tag me again if there's new content

**Comment by Vinnl** [comment]
OK, content should be final now @flodolo, with some strings changed in [7ccc4ac](https://github.com/mozilla/blurts-server/pull/2917/commits/7ccc4ac2ae38c6be84df3a03b5b0f76194e19b6f).

That also removes the distinction between US and non-US strings for the VPN recommendation, [as per Slack](https://mozilla.slack.com/archives/C0460QALW69/p1679590862194469?thread_ts=1679418777.468109&cid=C0460QALW69).

File: src/views/partials/breach-detail.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  const breachDetail = getBreachDetail(breach)
+  return `
+  <h2>${breachDetail.subhead}</h2>
+  <div>${breachDetail.body}</div>
+  `
+}
+
+export const breachDetailsPartial = data => `
+  <header class="breach-detail-header">
+    <img class="breach-detail-logo breach-logo" alt="" src="https://monitor.cdn.mozilla.net/img/logos/${data.breach.LogoPath}" />
```
Comment (pdehaan) [inline src/views/partials/breach-detail.js:132]: Q: can you confirm if this is correct. Are we still serving breach logos from the CDN or were you fetching and caching those locally now?

File: locales/en/app.ftl [hunk 10]
```diff
@@ -880,6 +874,8 @@ meta-desc = Find out if you’ve been part of a data breach with { -brand-fx-mon
 
 brand-fx-monitor = { -brand-fx-monitor }
 sign-in = Sign In
+header-nav-all-breaches = All breaches
+footer-nav-all-breaches = All breaches
```
[the remark this answers is not in the document]
Comment (toufali) [reply locales/en/app.ftl:878]: As is often the resolution between the 3 of us, let's just remove the header link altogether and pretend this never happened?

File: src/views/guestLayout.js [hunk 11]
```diff
@@ -61,6 +62,7 @@ const guestLayout = data => `
         <img src='/images/moz-logo-1color-white-rgb-01.svg' width='100' height='29' loading='lazy' alt='${getMessage('mozilla')}'>
       </a>
       <menu>
+        <li><a href='/breaches'>${getMessage('footer-nav-all-breaches')}</a></li>
         <li><a href='https://support.mozilla.org/kb/firefox-monitor-faq' target='_blank'>FAQ</a></li>
```
[the remark this answers is not in the document]
Comment (pdehaan) [reply src/views/guestLayout.js:66]: Actually, just noticed this while grepping for a different PR on `utm_*` parameters:
```sh
git grep -n "https://support.mozilla.org/kb/firefox-monitor-faq" src | cat

src/views/emails/email-2022.js:10:  faq: 'https://support.mozilla.org/kb/firefox-monitor-faq',
src/views/guestLayout.js:76:        <li><a href='https://support.mozilla.org/kb/firefox-monitor-faq' target='_blank'>FAQ</a></li>
src/views/mainLayout.js:105:        <li><a href='https://support.mozilla.org/kb/firefox-monitor-faq' target='_blank'>FAQ</a></li>
src/views/partials/landing.js:70:    <a href='https://support.mozilla.org/kb/firefox-monitor-faq' target='_blank'>${getMessage('see-all-faq')}</a>
```

Where `see-all-faq` string is: 
```sh
git grep -n "see-all-faq" locales/en/ | cat

locales/en/landing.ftl:40:see-all-faq = See all FAQs
```

So looks like "FAQ" is hardcoded 2&times; and uses localized "See all FAQs" elsewhere.
Not necessarily a bug, but just something to confirm possibly.

But maybe that makes sense since they are used in the landing page vs footer:
Screenshot 2023-03-23 at 4 17 10 PM

File: src/views/partials/breach-detail.js (hunk 9 above)
Comment (rhelmer) [reply src/views/partials/breach-detail.js:132]: Yes we're still using the CDN for this, these are larger company logos whereas the ones we fetch and cache locally are [favicons](https://en.wikipedia.org/wiki/Favicon).

We should definitely pick a lane here, I think the ideal would be if we start using a CDN for everything :) I'll file a ticket about this.

**Comment by Faby666** [comment]
Deja de espiarme!!!

File: src/utils/recommendations.js [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+export function getAllGenericRecommendations () {
+  return [
+    {
+      recommendationCopy: {
+        subhead: 'rec-gen-1-subhead',
+        body: 'rec-gen-1',
+        cta: 'rec-gen-1-cta'
+      },
+      ctaHref: 'https://monitor.firefox.com/security-tips#strong-passwords',
```
[the remark this answers is not in the document]
Comment (rhelmer) [reply src/utils/recommendations.js:15]: Yes we need to figure out what to do with /security-tips... I think we should just redirect to somewhere appropriate on mozilla.org or SUMO, I'll file an issue.

[the remark this answers is not in the document]
Comment (rhelmer) [reply src/utils/recommendations.js:15]: https://mozilla-hub.atlassian.net/browse/MNTOR-1464

File: src/views/partials/breach-detail.js (hunk 9 above)
Comment (rhelmer) [reply src/views/partials/breach-detail.js:132]: Filed https://mozilla-hub.atlassian.net/browse/MNTOR-1465

File: src/utils/recommendations.js [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+      weight: 91,
+      pathToGlyph: '/images/email-addresses',
+      recommendations: [
+        {
+          recommendationCopy: {
+            subhead: 'rec-email-mask-subhead',
+            body: 'rec-email',
+            cta: 'rec-email-cta'
+          },
+          ctaHref: 'https://relay.firefox.com/',
```
[the remark this answers is not in the document]
Comment (rhelmer) [reply src/utils/recommendations.js:217]: Separately I am working on attribution and analytics, I think we should tackle it holistically separately.

[the remark this answers is not in the document]
Comment (rhelmer) [reply src/utils/recommendations.js:217]: I'll work on this as part of the analytics work I'm already doing, thanks!

File: src/views/guestLayout.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (rhelmer) [reply src/views/guestLayout.js:66]: Oh hm good catch, I'm not sure if we generally translate FAQ but I think it's a reasonable thing to do? Since this landed in 54ad4b6d1 and not the current PR would you mind filing a ticket for this? Thanks!

File: src/utils/breach-logo.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * @param {unknkown} breach
```
[the remark this answers is not in the document]
Comment (rhelmer) [reply src/utils/breach-logo.js:6]: It looks like we just use `{object}` elsewhere ¯\_(ツ)_/¯ I can do that for consistency. It'd be nice to get stricter about type definitions, even if we don't go full-on Typescript we can still get a little bit of type checking and hints to our IDEs.

File: src/utils/breach-logo.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    '--orange-5',
+    '--yellow-5',
+    '--red-5',
+    '--pink-5'
+  ]
+
+  const charValue = name
+    .split('')
+    .map(letter => letter.codePointAt(0))
+    .reduce((sum, codePoint) => sum + codePoint)
```
[the remark this answers is not in the document]
Comment (rhelmer) [reply src/utils/breach-logo.js:37]: Yep optional per MDN :)

**Comment by toufali** [comment]
@flodolo minor string changes.  Possibly most impactful: https://github.com/mozilla/blurts-server/pull/2917/commits/72bf369fe67738d8ba5bbedca226a7d79eff60f8

I know we discussed caps before, but I think this is ok?  It's for a nav item in the footer.  "Terms & Privacy" is also in the footer and title-case 🙏

File: locales/en/app.ftl (hunk 10 above)
[the remark this answers is not in the document]
Comment (toufali) [reply locales/en/app.ftl]: Update: I removed the header link per UX, so this is non-issue. https://github.com/mozilla/blurts-server/pull/2917/commits/f1bdd6dd12f242cdcb889164b779fc75f7365987

File: src/client/css/partials/allBreachesPartial.css
[the remark this answers is not in the document]
Comment (toufali) [reply src/client/css/partials/allBreachesPartial.css:1]: Yes you're right, and there's a bit of work I need to do to make this clearer.  Unsurprisingly this will also likely land with code-splitting!  I updated to better conform with existing naming conventions here: https://github.com/mozilla/blurts-server/pull/2917/commits/ffdcec950d930ab85fca7ce35557398865aa8241

With code-splitting, part of the task will be to use camelCase, so you were ahead of schedule 🚀

File: src/client/js/partials/all-breaches.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const allBreachesPartial = document.querySelector('[data-partial="allBreachesPartial"]')
+
+if (allBreachesPartial) {
+  const search = document.getElementById('breach-search')
```
Comment (toufali) [inline src/client/js/partials/all-breaches.js:8]: micro-optimization nit!  Once we confirm `allBreachesPartial` we can henceforth query within that element instead of the entire document, e.g.  `allBreachesPartial.getElementById('breach-search')` and `allBreachesPartial.querySelectorAll('.breach-card')`. 

I'll see myself out.

**Review by toufali — APPROVED** [review]
I rush-pushed a bunch of stuff to get this over the finish line -- comments should be inline.  Thanks for the team effort! 🎉

File: src/client/css/global.css (hunk 5 above)
[the remark this answers is not in the document]
Comment (toufali) [reply src/client/css/global.css]: Per UX I removed the header "All Breaches" link – it was also a fast-solve for the user-menu bug 🙌 
https://github.com/mozilla/blurts-server/pull/2917/commits/f1bdd6dd12f242cdcb889164b779fc75f7365987

File: src/views/partials/breach-detail.js (hunk 9 above)
Comment (pdehaan) [reply src/views/partials/breach-detail.js:132]: Well, now I'm right confused.
I thought the current (v1) process is that @maxxcrawford lovingly creates a bespoke, artisanal 150&times;150 logo and submits a PR and somebody approves it and it gets hoisted into the CDN.
My understanding is that the new (v2) process did away with the CDN-ed, free-range images and only used the faviconed images (thus getting rid of the need for Maxx to make custom high quality logos).

I'm down with somebody finding a way to move the faviconed images into a single CDN solution (vs each machine in a cluster downloading and managing their own logos), but just wanted to confirm whether or not Maxx is still expected to toil away in the logo mines each time we need a logo.

Comment (toufali) [reply src/views/partials/breach-detail.js:132]: Ah yeah I assumed we were falling forward into the favicon automated route – I didn't have a chance to review these designs beforehand. I think we can release this now for the SEO, knowing Favicons on CDN is the way for future.

File: src/utils/recommendations.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (toufali) [reply src/utils/recommendations.js]: I couldn't let us go live with 404s so I removed the Security Tips links via https://github.com/mozilla/blurts-server/pull/2917/commits/e2dc855a3a385328fda7ef42bfc1d6002be8886d

File: src/utils/recommendations.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (toufali) [reply src/utils/recommendations.js]: I updated the UTMs for Relay and VPN via https://github.com/mozilla/blurts-server/pull/2917/commits/66fba1ac150f93fe78446b0a9477510d891cc33a

File: src/views/guestLayout.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (pdehaan) [reply src/views/guestLayout.js:65]: Done-ish. Filed as #2933. /cc @flodolo 

Since it doesn't look like y'all use GitHub issues very much, I'll copy/pasta it to Jira and update this comment with the Jira bug number since that seems like the source of truth(tm).

**Comment by toufali** [comment]
YOLO

File: src/utils/breach-logo.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/utils/breach-logo.js]: We don't have type definitions for anything, technically :) (Though that would be very high on my wish list.)

File: src/views/guestLayout.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (flodolo) [reply src/views/guestLayout.js:65]: (it should be translated, yes)

File: locales/en/app.ftl (hunk 10 above)
[the remark this answers is not in the document]
Comment (flodolo) [reply locales/en/app.ftl]: Footer/Header is an edge case, but in general it's better to have different strings for different contexts.
